### ISSUE TITLE
Local configuration feature

### ISSUE THREAD

**Issue by sergeyulyanov** [cereda/arara#34]
Hi Paolo!

A lot of programs that rely on configuration files, suggest several levels of settings detalization. For example, the program `latexmk` before processing checks the configuration files in a standard installation directory, in the user home directory and in the _current_ directory, from which it was started.

In my opinion it would be great if `arara` could also consider a local (e.g. project) configuration files. This can be useful in two cases: there is no need to extend the configuration file in the home directory by non reusable features specific only to a particular project and easy project portability to another platform with different settings.

As always I will be glad to know your opinion. Thank you!

**Comment by petobens** [comment]
Hi @cereda, this is not yet implemented in arara 4.0 right?
Thanks!

**Comment by cereda** [comment]
@petobens No, not at the moment. It looks quite tricky, I will see if I can make it work. `:)`

**Comment by petobens** [comment]
@cereda  there is no need to hurry, arara 4 is amazing as it stands.

**Comment by cereda** [comment]
This is the current workflow of `arara` 4.0:

1. First and foremost, `arara` will look for the existence of either `.araraconfig.yaml`, `araraconfig.yaml`, `.arararc.yaml` or `arararc.yaml` (in this order) in the user home directory (say, `/home/paulo`).

2. If found, the configuration is loaded and `arara` proceeds according to the settings.

3. Later, the command line arguments are parsed.

Item 3 is the point here: note that `arara` finds out the "reference file" (say, `/home/paulo/Documents/foo.tex`) as the last step, and the configuration is already loaded at this moment. I think this workflow is interesting because it gives more control on things, but I can see it complicates a potential local configuration feature.

Local configuration should be discovered based on the reference file path (say, `/home/paulo/Documents` + configuration file), but we will only have it when it's too late. To be honest, I'm not inclined into loading a configuration file twice because it looks clumsy to me (load global configuration, reset, load local configuration).

I thought of something which might be a compromise solution: exploit the `user.dir` system property which holds the user's current working directory. This value is directly available and I could include it as a prior verification ([it would be just a new `for` added](https://github.com/cereda/arara/blob/master/application/src/main/java/com/github/cereda/arara/utils/ConfigurationUtils.java#L101)).

But this solution has some caveats: it's about the current directory, which might differ from the reference file path. For example:

```
[paulo@cambridge /home/paulo/Documents] $ arara foo.tex
File reference: /home/paulo/Documents/foo.tex
User's current working directory: /home/paulo/Documents

[paulo@cambridge /home/paulo/Documents] $ arara bar/foo.tex
File reference: /home/paulo/Documents/bar/foo.tex
User's current working directory: /home/paulo/Documents

[paulo@cambridge /home/paulo/Documents] $ arara /home/paulo/foo.tex
File reference: /home/paulo/foo.tex
User's current working directory: /home/paulo/Documents
```

The clear advantage IMHO is that the implementation is very, very straightforward and it comes at virtually no cost and no significant changes to the code base.

Thoughts? `:)`

**Comment by cereda** [comment]
Let's see how it goes: I implemented the lookup for `user.dir` first then `user.home`. In other words, the user's working directory (`user.dir`) is inspected first for an occurrence of any of the configuration files and, if not found, the very same process is repeated for the user home directory (`user.home`). Quite straightforward implementation at almost no cost. `:)`

Feedback is surely welcome, so we can potentially (and finally) close this issue. :heart:

**Comment by alerque** [comment]
Feedback you say? WHY IS THIS NOT RELEASED? I'm pulling my hair out over here trying to figure out a way to hack around not having project-local configs. I'm using a continuous integration server and a runner to compile stuff and it really wants to keep stuff contained in the runner working directory. I don't have the option of writing stuff to the user dir (as that would run the risk of breaking other build runs) or the system. I went to review the source code to try to figure out a way to hack around this and discovered fix has been in place since March but not released! This release cycle is too long—release early and release often.

**Comment by alerque** [comment]
For anybody else that stumbles on this thread looking for a way to do this with arara 3.0, here is a hack for you. Arara asks Java what the user's home directory is and looks for stuff relative to there. You can trick java into returning this value as whatever you want by doing something like the following:

    export _JAVA_OPTIONS=-Duser.home=$(pwd)
    arara file.tex

**Comment by vermiculus** [comment]
@alerque Chill out, dude. Perhaps the release cycle should be more frequent, but yelling about it isn't constructive and definitely not the norm in this community. The release will be made when it is ready and no sooner. In the meantime, just use the development version.

That said, that's a clever little trick you've got :) Thanks for sharing.

**Comment by cereda** [comment]
Dear friend,

Thank you very much for your feedback. Also, thanks for sharing the trick of overriding the `user.home` system property. That said, I am sorry if I caused you any trouble, one way or another.

I know this might sound irrelevant for a technical issue, but I feel I should tell. I work two jobs and study. The only available time I have is between shifts. I also travel 600 km once or twice a week to attend classes and meetings. I try to keep up with my research when inside a bus or a metro, or late in the night, at home. The spare time is shared with my parents, my two cats, a few close friends and my own personal projects, including `arara`.

`arara` was released almost by accident. There was no intention on releasing it to the world, but a lot of friends pushed me into publishing it. The reception was surprisingly positive. But then a lot of bugs and nuisances emerged. I rewrote the tool almost a hundred times. Each time I fixed a bug, another two came to occupy the place of its predecessor.

I was about to give up. Really, I was near to the edge, close to throwing the towel. Then Marco, Brent and specially Nicola kindly supported me, which gave me the inspiration I needed to carry on. I then rewrote `arara` once again, this time, with different goals. The result was this current branch, version 4.0, which is far superior than the previous versions.

Sean, Ilhan, Christopher, Pedro, Andrew, a lot of friends came to the rescue. Each suggestion, improvement or fix helped pushing 4.0 to higher places. I believe we all have the greatest version ever, collaboratively built, but this statement comes at a cost. So many features, a lot of them under the hood, require a decent coverage. That is why a new user manual is required, completely rewritten from scratch. And this is the place we are now stuck.

I have a lot of trouble writing technical text, let alone in English. I have so much to say in the manual that I am afraid to be excessively verbose. I need to write sentences that are easy to grasp, yet precise and meaningful. Writing software is easy, writing good software is insanely hard. The very same analogy works for user manuals, I guess.

Again, I apologize for causing any trouble. I am trying really hard to find time to dedicate to this tool, but I have a lot of stuff going on in my life right now. I wish I could move `arara` to the top of my priorities, but that would certainly be unrealistic.

For now, I am sorry.

All the best, pal.

Paulo
